File: app/src/main/java/io/neoterm/view/ExtraKeysView.java
package io.neoterm.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.HapticFeedbackConstants;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;
import android.widget.ToggleButton;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import io.neoterm.customize.NeoTermPath;
import io.neoterm.customize.font.FontManager;
import io.neoterm.customize.eks.EksConfig;
import io.neoterm.customize.eks.EksConfigParser;
import io.neoterm.utils.FileUtils;
import io.neoterm.view.eks.ControlButton;
import io.neoterm.view.eks.ExtraButton;
import io.neoterm.view.eks.StatedControlButton;

import static io.neoterm.view.eks.ExtraButton.KEY_ARROW_DOWN;
import static io.neoterm.view.eks.ExtraButton.KEY_ARROW_LEFT;
import static io.neoterm.view.eks.ExtraButton.KEY_ARROW_RIGHT;
import static io.neoterm.view.eks.ExtraButton.KEY_ARROW_UP;
import static io.neoterm.view.eks.ExtraButton.KEY_CTRL;
import static io.neoterm.view.eks.ExtraButton.KEY_END;
import static io.neoterm.view.eks.ExtraButton.KEY_ESC;
import static io.neoterm.view.eks.ExtraButton.KEY_HOME;
import static io.neoterm.view.eks.ExtraButton.KEY_PAGE_DOWN;
import static io.neoterm.view.eks.ExtraButton.KEY_PAGE_UP;
import static io.neoterm.view.eks.ExtraButton.KEY_TAB;

/**
 * A view showing extra keys (such as Escape, Ctrl, Alt) not normally available on an Android soft
 * keyboard.
 */
public final class ExtraKeysView extends LinearLayout {

    public static final StatedControlButton CTRL = new StatedControlButton(KEY_CTRL);
    public static final ControlButton ESC = new ControlButton(KEY_ESC);
    public static final ControlButton TAB = new ControlButton(KEY_TAB);
    public static final ControlButton PAGE_UP = new ControlButton(KEY_PAGE_UP);
    public static final ControlButton PAGE_DOWN = new ControlButton(KEY_PAGE_DOWN);
    public static final ControlButton HOME = new ControlButton(KEY_HOME);
    public static final ControlButton END = new ControlButton(KEY_END);
    public static final ControlButton ARROW_UP = new ControlButton(KEY_ARROW_UP);
    public static final ControlButton ARROW_DOWN = new ControlButton(KEY_ARROW_DOWN);
    public static final ControlButton ARROW_LEFT = new ControlButton(KEY_ARROW_LEFT);
    public static final ControlButton ARROW_RIGHT = new ControlButton(KEY_ARROW_RIGHT);

    public static final String DEFAULT_FILE_CONTENT = "version " + EksConfigParser.PARSER_VERSION + "\n" +
            "program default\n" +
            "define - false\n" +
            "define / false\n" +
            "define | false\n";

    public static final int NORMAL_TEXT_COLOR = 0xFFFFFFFF;
    public static final int SELECTED_TEXT_COLOR = 0xFF80DEEA;

    private List<ExtraButton> builtinExtraKeys;
    private List<ExtraButton> userDefinedExtraKeys;

    private LinearLayout lineOne;
    private LinearLayout lineTwo;


    public ExtraKeysView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setGravity(Gravity.TOP);
        setOrientation(VERTICAL);
        HorizontalScrollView scrollOne = new HorizontalScrollView(context);
        HorizontalScrollView scrollTwo = new HorizontalScrollView(context);
        loadDefaultBuiltinExtraKeys();
        loadDefaultUserDefinedExtraKeys();
        lineOne = initLine(scrollOne);
        lineTwo = initLine(scrollTwo);
        addView(scrollOne);
        addView(scrollTwo);
        updateButtons();
    }

    private LinearLayout initLine(HorizontalScrollView scroll) {
        scroll.setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f));
        scroll.setFillViewport(true);
        scroll.setHorizontalScrollBarEnabled(false);
        LinearLayout line = new LinearLayout(getContext());
        line.setGravity(Gravity.START);
        line.setOrientation(LinearLayout.HORIZONTAL);
        scroll.addView(line);
        return line;
    }

    public boolean readControlButton() {
        return CTRL.readState();
    }

    public boolean readAltButton() {
        return false;
    }

    public void addUserDefinedButton(ExtraButton button) {
        addButton(userDefinedExtraKeys, button);
    }

    public void addBuiltinButton(ExtraButton button) {
        addButton(builtinExtraKeys, button);
    }

    private void addButton(List<ExtraButton> buttons, ExtraButton button) {
        if (!buttons.contains(button)) {
            buttons.add(button);
        }
    }

    public void clearUserDefinedButton() {
        userDefinedExtraKeys.clear();
    }

    public void loadDefaultUserDefinedExtraKeys() {
        userDefinedExtraKeys = new ArrayList<>(7);
        File defaultFile = new File(NeoTermPath.EKS_DEFAULT_FILE);
        if (!defaultFile.exists()) {
            generateDefaultFile(defaultFile);
        }

        clearUserDefinedButton();
        try {
            EksConfigParser parser = new EksConfigParser();
            parser.setInput(defaultFile);
            EksConfig config = parser.parse();
            userDefinedExtraKeys.addAll(config.getShortcutKeys());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void generateDefaultFile(File defaultFile) {
        FileUtils.INSTANCE.writeFile(defaultFile, DEFAULT_FILE_CONTENT.getBytes());
    }

    void loadDefaultBuiltinExtraKeys() {
        builtinExtraKeys = new ArrayList<>(7);
        builtinExtraKeys.clear();
        builtinExtraKeys.add(ESC);
        builtinExtraKeys.add(CTRL);
        builtinExtraKeys.add(TAB);
        builtinExtraKeys.add(ARROW_UP);
        builtinExtraKeys.add(ARROW_DOWN);
        builtinExtraKeys.add(ARROW_LEFT);
        builtinExtraKeys.add(ARROW_RIGHT);
        builtinExtraKeys.add(PAGE_UP);
        builtinExtraKeys.add(PAGE_DOWN);
        builtinExtraKeys.add(HOME);
        builtinExtraKeys.add(END);
    }

    public void updateButtons() {
        lineOne.removeAllViews();
        lineTwo.removeAllViews();
        for (final ExtraButton extraButton : userDefinedExtraKeys) {
            addExtraButton(lineOne, extraButton);
        }
        for (final ExtraButton extraButton : builtinExtraKeys) {
            addExtraButton(lineTwo, extraButton);
        }
    }

    private void addExtraButton(LinearLayout contentView, final ExtraButton extraButton) {
        final Button button;
        if (extraButton instanceof StatedControlButton) {
            StatedControlButton btn = ((StatedControlButton) extraButton);
            button = btn.toggleButton = new ToggleButton(getContext(), null, android.R.attr.buttonBarButtonStyle);
            button.setClickable(true);
            if (btn.initState) {
                btn.toggleButton.setChecked(true);
                btn.toggleButton.setTextColor(SELECTED_TEXT_COLOR);
            }
        } else {
            button = new Button(getContext(), null, android.R.attr.buttonBarButtonStyle);
        }

        button.setTypeface(FontManager.INSTANCE.getDefaultFont());
        button.setText(extraButton.buttonText);
        button.setTextColor(NORMAL_TEXT_COLOR);
        button.setAllCaps(false);

        button.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                button.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP);
                View root = getRootView();
                extraButton.onClick(root);
            }
        });
        contentView.addView(button);
    }

}

File: app/src/main/java/io/neoterm/view/ExtraKeysView.kt
package io.neoterm.view

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.Gravity
import android.view.HapticFeedbackConstants
import android.view.ViewGroup
import android.widget.Button
import android.widget.HorizontalScrollView
import android.widget.LinearLayout
import android.widget.ToggleButton

import java.io.File
import java.util.ArrayList

import io.neoterm.customize.eks.EksConfigParser
import io.neoterm.preference.NeoTermPath
import io.neoterm.utils.FileUtils
import io.neoterm.view.eks.ControlButton
import io.neoterm.view.eks.ExtraButton
import io.neoterm.view.eks.StatedControlButton

/**
 * A view showing extra keys (such as Escape, Ctrl, Alt) not normally available on an Android soft
 * keyboard.
 */
class ExtraKeysView(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs) {

    private var builtinExtraKeys: MutableList<ExtraButton>? = null
    private var userDefinedExtraKeys: MutableList<ExtraButton>? = null

    private val lineOne: LinearLayout
    private val lineTwo: LinearLayout
    private var typeface: Typeface? = null

    init {
        gravity = Gravity.TOP
        orientation = LinearLayout.VERTICAL
        val scrollOne = HorizontalScrollView(context)
        val scrollTwo = HorizontalScrollView(context)
        loadDefaultBuiltinExtraKeys()
        loadDefaultUserDefinedExtraKeys()
        lineOne = initLine(scrollOne)
        lineTwo = initLine(scrollTwo)
        addView(scrollOne)
        addView(scrollTwo)
        updateButtons()
    }

    private fun initLine(scroll: HorizontalScrollView): LinearLayout {
        scroll.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f)
        scroll.isFillViewport = true
        scroll.isHorizontalScrollBarEnabled = false
        val line = LinearLayout(context)
        line.gravity = Gravity.START
        line.orientation = LinearLayout.HORIZONTAL
        scroll.addView(line)
        return line
    }

    fun readControlButton(): Boolean {
        return CTRL.readState()
    }

    fun readAltButton(): Boolean {
        return ALT.readState()
    }

    fun addUserDefinedButton(button: ExtraButton) {
        addButton(userDefinedExtraKeys, button)
    }

    fun addBuiltinButton(button: ExtraButton) {
        addButton(builtinExtraKeys, button)
    }

    private fun addButton(buttons: MutableList<ExtraButton>?, button: ExtraButton) {
        if (buttons != null && !buttons.contains(button)) {
            buttons.add(button)
        }
    }

    fun clearUserDefinedButton() {
        userDefinedExtraKeys!!.clear()
    }

    fun loadDefaultUserDefinedExtraKeys() {
        userDefinedExtraKeys = ArrayList<ExtraButton>(7)
        val defaultFile = File(NeoTermPath.EKS_DEFAULT_FILE)
        if (!defaultFile.exists()) {
            generateDefaultFile(defaultFile)
        }

        clearUserDefinedButton()
        try {
            val parser = EksConfigParser()
            parser.setInput(defaultFile)
            val config = parser.parse()
            userDefinedExtraKeys!!.addAll(config.shortcutKeys)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun generateDefaultFile(defaultFile: File) {
        FileUtils.writeFile(defaultFile, DEFAULT_FILE_CONTENT.toByteArray())
    }

    internal fun loadDefaultBuiltinExtraKeys() {
        builtinExtraKeys = ArrayList<ExtraButton>(7)
        builtinExtraKeys!!.clear()
        builtinExtraKeys!!.add(ESC)
        builtinExtraKeys!!.add(CTRL)
        builtinExtraKeys!!.add(ALT)
        builtinExtraKeys!!.add(TAB)
        builtinExtraKeys!!.add(ARROW_UP)
        builtinExtraKeys!!.add(ARROW_DOWN)
        builtinExtraKeys!!.add(ARROW_LEFT)
        builtinExtraKeys!!.add(ARROW_RIGHT)
        builtinExtraKeys!!.add(PAGE_UP)
        builtinExtraKeys!!.add(PAGE_DOWN)
        builtinExtraKeys!!.add(HOME)
        builtinExtraKeys!!.add(END)
    }

    fun updateButtons() {
        lineOne.removeAllViews()
        lineTwo.removeAllViews()
        for (extraButton in userDefinedExtraKeys!!) {
            addExtraButton(lineOne, extraButton)
        }
        for (extraButton in builtinExtraKeys!!) {
            addExtraButton(lineTwo, extraButton)
        }
    }

    private fun addExtraButton(contentView: LinearLayout, extraButton: ExtraButton) {
        val button: Button
        if (extraButton is StatedControlButton) {
            val btn = extraButton
            val toggleButton = ToggleButton(context, null, android.R.attr.buttonBarButtonStyle)
            btn.toggleButton = toggleButton
            button = toggleButton

            button.setClickable(true)
            if (btn.initState) {
                btn.toggleButton!!.isChecked = true
                btn.toggleButton!!.setTextColor(SELECTED_TEXT_COLOR)
            }
        } else {
            button = Button(context, null, android.R.attr.buttonBarButtonStyle)
        }

        button.typeface = typeface
        button.text = extraButton.buttonText
        button.setTextColor(NORMAL_TEXT_COLOR)
        button.setAllCaps(false)

        button.setOnClickListener {
            button.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
            val root = rootView
            extraButton.onClick(root)
        }
        contentView.addView(button)
    }

    fun setTextColor(textColor: Int) {
        NORMAL_TEXT_COLOR = textColor
        updateButtons()
    }

    fun setTypeface(typeface: Typeface?) {
        this.typeface = typeface
        updateButtons()
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        val CTRL = StatedControlButton(ExtraButton.KEY_CTRL)
        @SuppressLint("StaticFieldLeak")
        val ALT = StatedControlButton(ExtraButton.KEY_ALT)
        val ESC = ControlButton(ExtraButton.KEY_ESC)
        val TAB = ControlButton(ExtraButton.KEY_TAB)
        val PAGE_UP = ControlButton(ExtraButton.KEY_PAGE_UP)
        val PAGE_DOWN = ControlButton(ExtraButton.KEY_PAGE_DOWN)
        val HOME = ControlButton(ExtraButton.KEY_HOME)
        val END = ControlButton(ExtraButton.KEY_END)
        val ARROW_UP = ControlButton(ExtraButton.KEY_ARROW_UP)
        val ARROW_DOWN = ControlButton(ExtraButton.KEY_ARROW_DOWN)
        val ARROW_LEFT = ControlButton(ExtraButton.KEY_ARROW_LEFT)
        val ARROW_RIGHT = ControlButton(ExtraButton.KEY_ARROW_RIGHT)

        val DEFAULT_FILE_CONTENT = "version " + EksConfigParser.PARSER_VERSION + "\n" +
                "program default\n" +
                "define - false\n" +
                "define / false\n" +
                "define | false\n" +
                "define $ false\n" +
                "define < false\n" +
                "define > false\n"

        var NORMAL_TEXT_COLOR = 0xFFFFFFFF.toInt()
        var SELECTED_TEXT_COLOR = 0xFF80DEEA.toInt()
    }
}
